### DISCUSSION TITLE
Expandable relationship fields

### DISCUSSION THREAD

**Discussion by jordantshaw** [fastapi/fastapi#9765] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
...
```

### Description

I would like to implement expandable fields in my endpoints.  For example, users could use a parameter called "expand", to pass a list of fields they would like to expand. These fields would be relationships to other tables.

For example the url: `fastapi.com/api/v1/customer`  would return something like the following.

```
{
    "customer_id": 1,
    "phone": {},
}
```

But the following url would expand the `phone` field and return the relationship. `fastapi.com/api/v1/customer?expand=phone`
```
{
    "customer_id": 1,
    "phone": {
        "id": 3,
        "phone_number": 123456789,
        "type": "cell"
    },
}
```

Any ideas on the best way to implement this in FastAPI.  

### Operating System

macOS

### FastAPI Version

0.92.0

### Python Version

Python 3.10.9

**Comment by jgould22** [comment]
This is relatively trivial with pydantic and optional types 

```python
class PhoneNumber(BaseModel):
     phone_number_id: int
     phone_number: str
     type: str

class Customer(BaseModel):
    customer_id: int
    phone: Optional[PhoneNumber] = None
 
```

Then if the client submits the expand attribute you just set the phone.

**Reply by jordantshaw** [reply]
Well it’s a little more complex than that I think. You can’t just return Null to the user as the actual value really could be null. I think you would need to set a default value of an empty dict untill the user expands that field. And that gets tricky.  

For example, how do you let the pydantic model know that the field is unset. They pydantic model will cause the ORM to lazy load that field if it is left unset. The alternative option is to specifically not load that relationshipfield, using ‘noload’.  But then that causes the values to be null.

**Reply by jgould22** [reply]
I would probably just inherit a model and have something like ExpandedCustomer (bad name but whatever) and not return those fields at all if they were not requested. It will however get real bad real fast if you have more than that one field and intend to have a lot of variants.

If you intend to do this with a bunch of fields then you are probably just going to end up making a haphazard implementation of graphql. If that is the case I would recommend you check out [strawberry](https://strawberry.rocks/docs/integrations/fastapi) which can [integrate with FastAPI](https://strawberry.rocks/docs/integrations/fastapi)

**Reply by jordantshaw** [reply]
Well I dont think it would be good design to just not return the field at all.  That causes the payload to change, and usually you want to return the same payload everytime.

**Reply by jgould22** [reply]
Sure, but the payload is changing either way.

I see three possible scenarios.

1 - The caller doesnt know about this expand param and just calls the customer endpoint, `phone` is included but and empty object -> gotta check the docs to see why that is

2 - They know what the docs say, they dont need the `phone` value and they call the endpoint and ignore that phone object -> the inclusion is not useful

3 - The want the phone, the know the query param and they call the endpoint with the expand param and inspect the phone value and whatever is contained there in.

The caller is forced to know that `phone` (and whatever else) would be included when they passed the expand param which means they have to look at the docs anyway.

If there are only really two variants here though (ie BaseCustomer and ExpandedCustomer) I would just put it on a different path with its own payload `/customer/{id}` and `/customer/{id}/details` or something.

**Reply by jordantshaw** [reply]
Field expansion is a fairly common architecture in REST APIs. Stripe, atlassian, etc... all support this without using graphql.  I would think there would be a fairly simple way to implement this in FastAPI, without have to alter the response model or have multiple endpoints.

**Reply by jgould22** [reply]
There very well might be and I just do not know how.

I would be interested to see what you find (or create) .

**Reply by jordantshaw** [reply]
I appreciate all the feedback, Ill update if I find an good solution.  Thanks for the help.

**Reply by jgould22** [reply]
Actually having looked at stripe and githubs api docs it looks like they dont seem to leave the object empty if it isnt expanded.

They put an id in its place, which makes way more sense to me.

```
{
    "customer_id": 1,
    "phone": 3
}
```

and 
```
{
    "customer_id": 1,
    "phone": {
        "id": 3,
        "phone_number": 123456789,
        "type": "cell"
    },
}
```
just becomes
```
class PhoneNumber(BaseModel):
     phone_number_id: int
     phone_number: str
     type: str

class Customer(BaseModel):
    customer_id: int
    phone: Union[int, PhoneNumber]
```

no?

**Reply by jordantshaw** [reply]
Yeah, that makes sense, sort of.  However the implementation is still not clear.  Because now you can either load the phone_id or the phone object.  How do you determine that from the DB side?  the big issue Im facing is with lazy loading.  Because the response model contains "phone", even if I dont include the phone relationship in the ORM query, it still gets loaded, because the response model causes the ORM to lazy load the phone object.  I can set this load to to be "noload", but then the phone object just returns as "null" instead of unset, and im back to the question of how do I let the model know its unset, instead of null.

Also, not sure how well that would work if you had an array of of objects.  As you would still have to do the join to the child table to get the IDs, and that doesnt really help solve anything.  The objective with the expand parameter is to avoid joining to child tables unless the user requests it.

**Reply by karimelhajoui63** [reply]
Hi, I'm looking for the exact same thing 
Do you ended up with a viable solution @jordantshaw? :)

**Comment by YuriiMotov** [comment]
Similar: https://github.com/fastapi/fastapi/discussions/13404
